src/capture/capture_video_ffmpeg.cpp

```diff
@@ -0,0 +1,733 @@
+	           video_encoder.frames_per_second != rounded_fps) {
+
+		CaptureVideoFinalise();
+		if (!video_encoder.Init(image.params.width,
+		                        image.params.height,
+		                        image.params.pixel_format,
+		                        rounded_fps)) {
+			LOG_ERR("FFMPEG: Video encoder failed to init");
+			video_encoder.Free();
+			return;
+		}
+
+		mutex.lock();
+		video_encoder.is_initalised = true;
+		mutex.unlock();
+		waiter.notify_all();
+	}
+
+	RenderedImage copy = image.deep_copy();
+	if (!video_encoder.queue.Enqueue(std::move(copy))) {
+		// Should never happen but avoid a memory leak if it does
+		LOG_ERR("FFMPEG: Video encoder queue is stopped");
+		copy.free();
+	}
+}
+
+void FfmpegEncoder::CaptureVideoAddAudioData(const uint32_t sample_rate,
+                                                 const uint32_t num_sample_frames,
+                                                 const int16_t* sample_frames)
+{
+	if (!audio_encoder.is_initalised) {
+		if (!audio_encoder.Init(sample_rate)) {
+			LOG_ERR("FFMPEG: Failed to init audio encoder");
+			audio_encoder.Free();
+			return;
+		}
+
+		bool init_muxer = false;
+		if (!muxer.is_initalised && video_encoder.is_initalised) {
+			init_muxer = muxer.Init(video_encoder, audio_encoder);
+			if (!init_muxer) {
+				LOG_ERR("FFMPEG: Failed to init muxer");
+				muxer.Free();
+				audio_encoder.Free();
+				return;
+			}
+		}
+
+		mutex.lock();
+		audio_encoder.is_initalised = true;
+		muxer.is_initalised         = muxer.is_initalised || init_muxer;
+		mutex.unlock();
+		waiter.notify_all();
+
+	} else if (audio_encoder.sample_rate != sample_rate) {
+		CaptureVideoFinalise();
+		if (!audio_encoder.Init(sample_rate)) {
+			LOG_ERR("FFMPEG: Failed to init audio encoder");
+			audio_encoder.Free();
+			return;
+		}
+
+		mutex.lock();
+		audio_encoder.is_initalised = true;
+		mutex.unlock();
+		waiter.notify_all();
+	}
+
+	const size_t num_samples = num_sample_frames * SamplesPerFrame;
+	std::vector<int16_t> audio_data(sample_frames, sample_frames + num_samples);
+	audio_encoder.queue.BulkEnqueue(audio_data, num_samples);
+}
+
+void FfmpegEncoder::StopQueues()
+{
+	std::unique_lock<std::mutex> lock(mutex);
+
+	// Set these first so none of the threads start another iteration
+	// Before the encoders get re-initalised
+	audio_encoder.is_initalised = false;
+	video_encoder.is_initalised = false;
+	muxer.is_initalised         = false;
+
+	audio_encoder.queue.Stop();
+	video_encoder.queue.Stop();
+	while (audio_encoder.is_working || video_encoder.is_working) {
+		waiter.wait(lock);
+	}
+	muxer.queue.Stop();
+	while (muxer.is_working) {
+		waiter.wait(lock);
+	}
+
+	// At least the muxer queue has been measured
+	// to have stale items left in it if it is never started
+	// due to one of the encoders failing to init.
+	// Explicitly clear the queues out so we don't have to worry
+	// about stale work with incorrect video/audio specs for the next recording.
```

Comment:
Minor: break text properly

Reply:
I re-ran the format script.  It always has its way with my comment blocks trying to limit line size.  I don't know how you want me to "properly" wrap it but see if it looks better now?